llvm/include/llvm/Object/SYCLBIN.h

```diff
@@ -0,0 +1,130 @@
+//===- SYCLBIN.h - SYCLBIN binary format support ----------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_OBJECT_SYCLBIN_H
+#define LLVM_OBJECT_SYCLBIN_H
+
+#include "llvm/ADT/SmallString.h"
+#include "llvm/SYCLPostLink/ModuleSplitter.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include <string>
+
+namespace llvm {
+
+namespace object {
+
+// Representation of a SYCLBIN binary object. This is intended for use as an
+// image inside a OffloadBinary.
+class SYCLBIN {
```

Comment:
Is it possible to encode all information from SYCLBIN in `StringData` provided by [OffloadBinary.h](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Object/OffloadBinary.h#L61) ?

It would significantly improve chances of upstreaming of this. Also it simplifies the maintainability. I understand that it would lead to a loss of much your effort. However, it looks like a good investment in overall.

`StringData` allows to encode string and binary information. So everything could be put there accompanied by some key names like: "Properties".

If you say that runtime is capable to reuse the LLVM's implementation of OffloadBinary, then it shouldn't be so difficult to just read values from `StringData`.

Reply:
I am not sure I comprehend what parts should be moved to the `StringData` and what should be moved to `Image`? The intention is to have multi-layered file format and I don't see how embedding parts of this into StringData over just having it encoded in the image data.

Feel free to ping me offline and we can set up a meeting. This seems like something that will be complex to align on over text.

Reply:
My initial comment was misleading.